File: historical_camera/ios/Runner/HistoricalCamera/FilterParams.swift
import Foundation

/// Filter parameters received from Dart (docs/02 §2).
///
/// Field order matches the Dart declaration; the shader uniform order in
/// `FilterUniforms` follows it as well. Defaults are the documented neutral
/// values — zero-initialization is forbidden (docs/02 §2).
struct FilterParams {
    var monochrome: Float = 0
    var sepia: Float = 0
    var saturation: Float = 1
    var contrast: Float = 1
    var brightness: Float = 0
    var warmth: Float = 0
    var fade: Float = 0
    var grain: Float = 0
    var grainSize: Float = 1
    var vignette: Float = 0
    var scratches: Float = 0
    var dust: Float = 0
    var jitter: Float = 0
    var halation: Float = 0
    var blur: Float = 0
    var orthochromatic: Float = 0
    var engraving: Float = 0
    var hatchScale: Float = 1
    var inkPainting: Float = 0
    var paperTexture: Float = 0

    static let neutral = FilterParams()

    init() {}

    /// Parses the channel map; missing keys keep their neutral value.
    init(map: [String: Any]) {
        func value(_ key: String, _ fallback: Float) -> Float {
            (map[key] as? NSNumber)?.floatValue ?? fallback
        }
        monochrome = value("monochrome", 0)
        sepia = value("sepia", 0)
        saturation = value("saturation", 1)
        contrast = value("contrast", 1)
        brightness = value("brightness", 0)
        warmth = value("warmth", 0)
        fade = value("fade", 0)
        grain = value("grain", 0)
        grainSize = value("grainSize", 1)
        vignette = value("vignette", 0)
        scratches = value("scratches", 0)
        dust = value("dust", 0)
        jitter = value("jitter", 0)
        halation = value("halation", 0)
        blur = value("blur", 0)
        orthochromatic = value("orthochromatic", 0)
        engraving = value("engraving", 0)
        hatchScale = value("hatchScale", 1)
        inkPainting = value("inkPainting", 0)
        paperTexture = value("paperTexture", 0)
    }
}

File: historical_camera/android/app/src/main/kotlin/com/daisuke7/historical/camera/historicalcamera/FilterParams.kt
package com.daisuke7.historical.camera.historicalcamera

/**
 * Filter parameters received from Dart (docs/02 §2).
 *
 * Field order matches the Dart declaration; defaults are the documented
 * neutral values — zero-initialization is forbidden (docs/02 §2).
 */
data class FilterParams(
    val monochrome: Float = 0f,
    val sepia: Float = 0f,
    val saturation: Float = 1f,
    val contrast: Float = 1f,
    val brightness: Float = 0f,
    val warmth: Float = 0f,
    val fade: Float = 0f,
    val grain: Float = 0f,
    val grainSize: Float = 1f,
    val vignette: Float = 0f,
    val scratches: Float = 0f,
    val dust: Float = 0f,
    val jitter: Float = 0f,
    val halation: Float = 0f,
    val blur: Float = 0f,
    val orthochromatic: Float = 0f,
    val engraving: Float = 0f,
    val hatchScale: Float = 1f,
    val inkPainting: Float = 0f,
    val paperTexture: Float = 0f,
) {
    /** Values in the docs/02 §2 declaration order, for `uniform float[20]`. */
    fun toFloatArray(): FloatArray = floatArrayOf(
        monochrome, sepia, saturation, contrast, brightness, warmth, fade,
        grain, grainSize, vignette, scratches, dust, jitter, halation, blur,
        orthochromatic, engraving, hatchScale, inkPainting, paperTexture,
    )

    companion object {
        val NEUTRAL = FilterParams()

        /** Channel values arrive as Number (docs/02 §2 type contract). */
        fun fromMap(map: Map<*, *>): FilterParams {
            fun f(key: String, fallback: Float): Float =
                (map[key] as? Number)?.toFloat() ?: fallback
            return FilterParams(
                monochrome = f("monochrome", 0f),
                sepia = f("sepia", 0f),
                saturation = f("saturation", 1f),
                contrast = f("contrast", 1f),
                brightness = f("brightness", 0f),
                warmth = f("warmth", 0f),
                fade = f("fade", 0f),
                grain = f("grain", 0f),
                grainSize = f("grainSize", 1f),
                vignette = f("vignette", 0f),
                scratches = f("scratches", 0f),
                dust = f("dust", 0f),
                jitter = f("jitter", 0f),
                halation = f("halation", 0f),
                blur = f("blur", 0f),
                orthochromatic = f("orthochromatic", 0f),
                engraving = f("engraving", 0f),
                hatchScale = f("hatchScale", 1f),
                inkPainting = f("inkPainting", 0f),
                paperTexture = f("paperTexture", 0f),
            )
        }
    }
}
